rm problematic variables under lambdas and evars for evar instantiation

Tragicus opened the pull request:
<!-- Thank you for your contribution.
     Make sure you read the contributing guide and fill this template. -->
The unification algorithm sometimes fails on problems of the form `?a = t`  because variables that are not in the scope of `?a` appear in `t`. There are two cases where we can easily solve the problem, namely the one where the variable appears in a beta-redex, as in `(fun _ => t) x`, and the one where the variable appears in an argument of an evar, e.g. `?b x`, since we can instantiate the evar to forget the argument.
For instance, unifying `forall x : ?T, f (?a x)` and `forall _ : ?T, ?b` reduces to unifying `f (?a x)` and `?b`, which fails because `x` is not available when instantiating `?b`.
This PR adds these cases in `pretyping/evarsolve`.

<!-- If this is a bug fix, make sure the bug was reported beforehand. -->
Fixes / closes #????


<!-- Remove anything that doesn't apply in the following checklist. -->

<!-- If there is a user-visible change and testing is not prohibitively expensive: -->
- [ ] Added / updated **test-suite**.

<!-- If this is a feature pull request / breaks compatibility: -->
- [ ] Added **changelog**.
- [ ] Added / updated **documentation**.
  <!-- Check if the following applies, otherwise remove these lines. -->
  - [ ] Documented any new / changed **user messages**.
  - [ ] Updated **documented syntax** by running `make doc_gram_rsts`.

<!-- If this breaks external libraries or plugins in CI: -->
- [ ] Opened **overlay** pull requests.

<!-- Pointers to relevant developer documentation:

Contributing guide: https://github.com/coq/coq/blob/master/CONTRIBUTING.md

Test-suite: https://github.com/coq/coq/blob/master/test-suite/README.md

Changelog: https://github.com/coq/coq/blob/master/doc/changelog/README.md

Building the doc: https://github.com/coq/coq/blob/master/doc/README.md
Sphinx: https://github.com/coq/coq/blob/master/doc/sphinx/README.rst
doc_gram: https://github.com/coq/coq/blob/master/doc/tools/docgram/README.md

Overlays: https://github.com/coq/coq/blob/master/dev/ci/user-overlays/README.md

SkySkimmer commented:
@coqbot run full c

SkySkimmer commented:
@coqbot run full ci

ppedrot commented:
@coqbot ci minimize

JasonGross commented:
bug minimizer should work on ci-iris now with https://github.com/coq-community/run-coq-bug-minimizer/commit/5b6d8188224f88af334c27241d273ad06fd12ec4: @coqbot ci minimize ci-iris

JasonGross commented:
oops, sorry, let's try again.  @coqbot ci minimize ci-iris

Tragicus commented:
I can not get the minimized file from `unimath` to compile. Regarding the issue in `fiat-parser`, `eauto` applies `iffR_pull` to the goal `iffR _ _` by setting `P` and `Q` as constant functions and the next argument as the identity, which is correct, so I do not quite know what to say.

ppedrot commented:
@Tragicus Did you set the flags in the header accordingly to try to compile the unimath snippet? Especially `noinit -indices-matter -type-in-type`.

Tragicus commented:
Ah, no, I did not, although there was a typing error in the definition of `hProp` and I could not convince myself that it should work, so I am not too hopeful. But thank you anyway, I will try again.

Tragicus commented:
There is quite the same issue in the `argosy` example. `eauto` finds that the identity is a solution of `forall x : ?T, f (?e x)`, which entails dropping the argument of `?e` defining `?T` as `f ?e`. This of course breaks subsequent goals.

Tragicus commented:
The `unimath` file does compile with the options. The issue is that there is a `refine bla` where some subterm of `bla` is a lambda with holes, whose type is asserted to be non-dependent by the context. `bla` does not typecheck on `master`, while it does with this branch. `refine bla` still succeeds on master, whereas on this branch the hole in the lambda is partially instantiated to forget the bound variable, which leads to an error. I also noted that in the types, the variable appears only in the third argument of `abs_ax` which is defined as `fun T n _ => T n`.

JasonGross commented:
So it sounds like many of the issues are "eauto got stronger".  Are the proofs easily fixable?

> `bla` does not typecheck on `master`, while it does with this branch. `refine bla` still succeeds on master

This is confusing.  How does `refine bla` succeed without wrecking `Qed` if `bla` is ill-typed?  Or do you just mean that `bla` only typechecks with BIDI type hints from the goal?

Tragicus commented:
I think so, but as I started debugging with hott who has typeclasses issues, I have not tried yet.
To be precise, `bla` is `(pattern)` and `refine '(pattern)` succeeds whereas `refine (pattern)` fails. Also, technically, the term is well-typed, but only because the bound variable disappears when you reduce and the unification algorithm does not delta-reduce when instantating evars, so it does not see it. So `Check bla` fails.

JasonGross reviewed (changes requested):
> There are two cases where we can easily solve the problem, namely the one where the variable appears in a beta-redex, as in `(fun _ => t) x`, and the one where the variable appears in an argument of an evar, e.g. `?b x`, since we can instantiate the evar to forget the argument.

The beta-redex case seems unambiguously good (if you want to split that off into its own PR, it will probably not break the CI and can probably be merged more quickly).

The evar restriction case, however, is not unique unless the evar is irreducibly in a head position (I don't know the proper terminology for this).  If some other reduction (for example, beta) can remove the evar from the term, then we don't necessarily want to restrict its context.  If you want to restrict evar restriction to the case where the evar being restricted is the head of the term, or (recursively) the head of the body of a lambda/for all or an argument to a constructor, (co)inductive type family, match, or Axiom, I would be in favor. 

I would also support making unification more customizable, and guarding this (and other non-unique) behavior with a flag (e.g., `Set Unification Heuristic Evar Restriction` or something).

Also curious what other devs involved in unification (cc @herbelin @SkySkimmer ) think

Tragicus commented:
Unfortunately, the variables that appear as arguments of evars are the main issue. I do see the example I give in [the description of the PR](https://github.com/coq/coq/pull/19822#issue-2645923952), where `f` is opaque (or more precisely a projection), so the problem does not reduce (and we anyway do not want to delta-reduce).

JasonGross commented:
What are your thoughts then on guarding the behavior by `Set Unification Heuristic Evar Restriction` or similar?  It seems fine to have it on by default, and disable it in the compat files & in the CI devs that are not otherwise easy to fix.

Tragicus commented:
Considering that https://github.com/coq/coq/pull/19833 also fails and that it looks like there is the same typeclass issue in hott, we should first see what happens when things are fixed there. Maybe we will see that the applied evar case only triggers `eauto` to sometimes do too much and it may be easy to fix without introducing more options. I will also try removing the beta-redex case from here, just to see what happens.

JasonGross commented:
@coqbot run full ci

Tragicus commented:
The error in the `hott` example looks like there is a `repeat split` at the end where some `split` succeeds while leaving out some unification constraints. Rocq fails to solve them outside of the `repeat`, so the error is not caught and the whole line fails.

JasonGross commented:
> The error in the `hott` example looks like there is a `repeat split` at the end where some `split` succeeds while leaving out some unification constraints. Rocq fails to solve them outside of the `repeat`, so the error is not caught and the whole line fails.

@Tragicus could you minimize this to a smaller example that displays the change in behavior?  I presume that the issue is something like that we have `Class Foo := bar : forall x y, baz = qux` and `split` on a goal of type `Foo` would previously fail, resulting in typeclass resolution picking up the goal, whereas now `split` tries to apply `eq_refl` which succeeds but leaves over unsolved constraints?

Plausibly tactics that have both an `e` variant and a non-`e` variant (`split` vs `esplit`, `apply` vs `eapply`) should call `solve_unification_constraints` on the non-`e` variant?  I think this might lead to more predictable behavior overall, though I'm not sure.

Tragicus commented:
I guess given the name of the job and the message trace that `ci-elpi_test` does the tests in `coq-elpi`. However, I can not find any `test_quotation.v` in the repo. Where should I look at?

Tragicus commented:
> > The error in the `hott` example looks like there is a `repeat split` at the end where some `split` succeeds while leaving out some unification constraints. Rocq fails to solve them outside of the `repeat`, so the error is not caught and the whole line fails.
> 
> @Tragicus could you minimize this to a smaller example that displays the change in behavior? I presume that the issue is something like that we have `Class Foo := bar : forall x y, baz = qux` and `split` on a goal of type `Foo` would previously fail, resulting in typeclass resolution picking up the goal, whereas now `split` tries to apply `eq_refl` which succeeds but leaves over unsolved constraints?
> 
> Plausibly tactics that have both an `e` variant and a non-`e` variant (`split` vs `esplit`, `apply` vs `eapply`) should call `solve_unification_constraints` on the non-`e` variant? I think this might lead to more predictable behavior overall, though I'm not sure.

Here is the minimized version of the minimized `hott` example:
```coq
Declare ML Module "ltac_plugin:coq-core.plugins.ltac".

Class SgOp A := sg_op: forall (_ _ : A), A.
Class MonUnit A := mon_unit: A.
Class Negate A := negate: forall _ : A, A.

Class RightInverse {A} {B} {C} (op : forall (_ : A) (_ : B), C) (inv : forall _ : A, B) (unit : C)
  := right_inverse: forall x, op x (inv x) = unit.



Class IsGroup A {Aop : SgOp A} {Aunit : MonUnit A} {Anegate : Negate A} :=
  { negate_r : RightInverse sg_op negate mon_unit }.

Record Group := {
  group_type :> Type;
  group_sgop :: SgOp group_type;
  group_unit :: MonUnit group_type;
  group_inverse :: Negate group_type;
  group_isgroup :: IsGroup group_type;
}.


Section FreeProduct.

Context (G H K : Group).
Definition amal_type : Type.
Admitted.

Definition AmalgamatedFreeProduct : Group.
Proof.
  simple notypeclasses refine (Build_Group amal_type _ _ _ _).
  4: split.
  4: split.
```

Tragicus commented:
The behavior is indeed different, although the difference is not the same, here it simply solves the goal.

JasonGross commented:
Do you want to make a PR to Coq that replaces all non-e-tactics with an e- variant (so, like, `split`, `exists`, probably a couple others) with `tactic; solve_constraints`?  If that passes the CI, I'd support merging it, and I suspect it'll solve some of the compat issues here (at least HoTT, if not others).  And the minimized HoTT example (and probably also the extra minimized example I gave) should be added to the test-suite, either here or in that PR.

Tragicus commented:
I am a bit unsure about that. I tried adding a `Refine.solve_constraints` in `tactics/tactics.ml` at the end of `general_apply` and it solved the issue., but I do not know this part of the code, so I am not sure what I should do precisely. Regarding defining `tac` as `etac; solve_constraints`, I thought allowing the `etac`s to introduce new evars made then very unstable, so doing what you propose would also make the non-e-`tac`s unstable, whereas `tac` leaving unsolved unification constraints around looks more like a small bug.

JasonGross commented:
Ah, I am not suggesting that we define `tac` as `etac; solve_constraints`, merely that for all tactics where we have both `tac` and `etac`, `tac` should be followed by `solve_constraints`.  That is, non-`e`-tactics should neither introduce new evars nor modify evars in ways that require delayed constraints.



> I tried adding a `Refine.solve_constraints` in `tactics/tactics.ml` at the end of `general_apply` and it solved the issue.

Let's cc @coq/tactics-maintainers ?  I don't know this part of the code well either

SkySkimmer commented:
It may make sense to put it in https://github.com/coq/coq/blob/da4e34bea12151a8e78d9b60b4fc8f263be6c1f3/tactics/tacticals.ml#L406 
I think most tactics that have e and non-e variants go through one of the WITHHOLES tacticals which use this code in the non-e case

Tragicus commented:
This is a tactic, so I put it in `tclMADDELAYEDWITHHOLES` and `with_holes_check` in place on the `tclUNIT ()` after the calls to `check_evars`. `hott` still does not compile, though.

SkySkimmer commented:
maybe better before the check_evars instead of after, although I don't know if that would change the result on hott

Tragicus commented:
It does not work either.

SkySkimmer commented:
It seems that the iris issue is caused by `?x == ?y a` being solved by `?x := ?y a` instead of `?y := ?x`, where the later would `inherit_evar_flags` but the new behaviour doesn't. `?x` is shelved prealably and the lack of evar aliasing info means `?y` doesn't end up shelved.

Tragicus commented:
I added something in `evarsolve/instantiate_evar` to shelve the new variable when we instantiate on old shelved variable with it. However, this feels like it was doing the right thing and it is just that we should instantiate the new variable with the old one and not the converse.

JasonGross commented:
> I added something in `evarsolve/instantiate_evar` to shelve the new variable when we instantiate on old shelved variable with it.

This seems wrong (as you say?)?  If we have (untested)
```coq
Goal forall A : Prop, (exists x : A, True) /\ A.
simple refine (fun A => conj _ ?[a]).
epose _ as x.
exists x.
instantiate (1:=?a) in (value of x).
```
the final line should not shelve the second goal.  This should probably be true whether we go through instantiate or unification, though I'm not 100% sure about that.

Tragicus commented:
Can I see the rest of the CI?

JasonGross commented:
@coqbot run full ci

SkySkimmer commented:
You can reopen if you can make progress
